File: C/prime_numbers.c
/**
 * Prime Number Checker - Multiple Methods
 * A prime number is a natural number greater than 1 that has no positive divisors other than 1 and itself.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

/**
 * Method 1: Basic approach - check all numbers up to n
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
int isPrimeBasic(int n) {
    if (n <= 1) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    
    for (int i = 3; i < n; i += 2) {
        if (n % i == 0) return 0;
    }
    return 1;
}

/**
 * Method 2: Optimized - check only up to sqrt(n)
 * Time Complexity: O(sqrt(n))
 * Space Complexity: O(1)
 */
int isPrimeOptimized(int n) {
    if (n <= 1) return 0;
    if (n <= 3) return 1;
    if (n % 2 == 0 || n % 3 == 0) return 0;
    
    for (int i = 5; i * i <= n; i += 6) {
        if (n % i == 0 || n % (i + 2) == 0) return 0;
    }
    return 1;
}

/**
 * Method 3: Using square root
 * Time Complexity: O(sqrt(n))
 * Space Complexity: O(1)
 */
int isPrimeSqrt(int n) {
    if (n <= 1) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    
    int sqrtN = (int)sqrt(n);
    for (int i = 3; i <= sqrtN; i += 2) {
        if (n % i == 0) return 0;
    }
    return 1;
}

/**
 * Method 4: Sieve of Eratosthenes (for generating multiple primes)
 * Time Complexity: O(n log log n)
 * Space Complexity: O(n)
 */
void generatePrimesSieve(int limit, int primes[], int *count) {
    if (limit < 2) {
        *count = 0;
        return;
    }
    
    int *isPrime = (int*)malloc((limit + 1) * sizeof(int));
    for (int i = 0; i <= limit; i++) {
        isPrime[i] = 1;
    }
    isPrime[0] = isPrime[1] = 0;
    
    for (int i = 2; i * i <= limit; i++) {
        if (isPrime[i]) {
            for (int j = i * i; j <= limit; j += i) {
                isPrime[j] = 0;
            }
        }
    }
    
    *count = 0;
    for (int i = 2; i <= limit; i++) {
        if (isPrime[i]) {
            primes[(*count)++] = i;
        }
    }
    
    free(isPrime);
}

/**
 * Method 5: Generate first n prime numbers
 */
void firstNPrimes(int n, int primes[]) {
    int count = 0;
    int num = 2;
    
    while (count < n) {
        if (isPrimeOptimized(num)) {
            primes[count++] = num;
        }
        num++;
    }
}

// Helper function to print array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Test examples
int main() {
    printf("Prime Number Checker - Testing All Methods\n\n");
    printf("============================================================\n");
    
    int testNumbers[] = {1, 2, 3, 4, 5, 10, 11, 17, 20, 29, 30, 97, 100};
    int numTests = sizeof(testNumbers) / sizeof(testNumbers[0]);
    
    printf("Individual Number Tests:\n");
    printf("------------------------------------------------------------\n");
    for (int i = 0; i < numTests; i++) {
        int num = testNumbers[i];
        printf("Number %3d: Method 1: %s, Method 2: %s, Method 3: %s\n",
            num, 
            isPrimeBasic(num) ? "true " : "false",
            isPrimeOptimized(num) ? "true " : "false",
            isPrimeSqrt(num) ? "true " : "false");
    }
    
    printf("\n============================================================\n");
    printf("Sieve of Eratosthenes - Prime numbers up to 50:\n");
    int primes[1000];
    int count;
    generatePrimesSieve(50, primes, &count);
    printArray(primes, count);
    
    printf("\n============================================================\n");
    printf("First 15 prime numbers:\n");
    firstNPrimes(15, primes);
    printArray(primes, 15);
    
    return 0;
}

File: Python/prime_numbers.py
"""
Prime Number Checker - Multiple Methods
A prime number is a natural number greater than 1 that has no positive divisors other than 1 and itself.
"""


def is_prime_basic(n):
    """
    Method 1: Basic approach - check all numbers up to n
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n <= 1:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    for i in range(3, n, 2):
        if n % i == 0:
            return False
    
    return True


def is_prime_optimized(n):
    """
    Method 2: Optimized - check only up to sqrt(n)
    Time Complexity: O(sqrt(n))
    Space Complexity: O(1)
    """
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    
    return True


def is_prime_sqrt(n):
    """
    Method 3: Using square root
    Time Complexity: O(sqrt(n))
    Space Complexity: O(1)
    """
    import math
    
    if n <= 1:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    sqrt_n = int(math.sqrt(n))
    for i in range(3, sqrt_n + 1, 2):
        if n % i == 0:
            return False
    
    return True


def generate_primes_sieve(limit):
    """
    Method 4: Sieve of Eratosthenes (for generating multiple primes)
    Time Complexity: O(n log log n)
    Space Complexity: O(n)
    """
    if limit < 2:
        return []
    
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False
    
    for i in range(2, int(limit**0.5) + 1):
        if is_prime[i]:
            for j in range(i*i, limit + 1, i):
                is_prime[j] = False
    
    return [num for num in range(limit + 1) if is_prime[num]]


def first_n_primes(n):
    """
    Method 5: Generate first n prime numbers
    """
    primes = []
    num = 2
    
    while len(primes) < n:
        if is_prime_optimized(num):
            primes.append(num)
        num += 1
    
    return primes


# Test examples
if __name__ == "__main__":
    print("Prime Number Checker - Testing All Methods\n")
    print("=" * 60)
    
    test_numbers = [1, 2, 3, 4, 5, 10, 11, 17, 20, 29, 30, 97, 100]
    
    print("Individual Number Tests:")
    print("-" * 60)
    for num in test_numbers:
        print(f"Number {num:3d}: Method 1: {is_prime_basic(num)}, "
              f"Method 2: {is_prime_optimized(num)}, "
              f"Method 3: {is_prime_sqrt(num)}")
    
    print("\n" + "=" * 60)
    print("Sieve of Eratosthenes - Prime numbers up to 50:")
    print(generate_primes_sieve(50))
    
    print("\n" + "=" * 60)
    print("First 15 prime numbers:")
    print(first_n_primes(15))
